lib/outh/google-outh.ts

```diff
@@ -0,0 +1,106 @@
+  Promise.resolve().then(_ => callback.apply(null, [...arguments].slice(1)));
+  return 0;
+};
+
+// If modifying these this.scopes, delete your previously saved credentials
+// at ~/.credentials/sheets.googleapis.com-nodejs-pwmetrics.json
+const SCOPES: Array<string> = ['https://www.googleapis.com/auth/spreadsheets'];
+const EEXIST: string = 'EEXIST';
+
+class GoogleOuth {
+  private tokenDir: string = path.join((process.env.HOME || process.env.HOMEPATH || process.env.USERPROFILE), '/.credentials/');
+  private tokenPath: string = path.join(this.tokenDir, 'sheets.googleapis.com-nodejs-pwmetrics.json');
+
+  async authenticate(clientSecret:AuthorizeCredentials): Promise<Oauth2Client> {
+    try {
+      return await this.authorize(clientSecret);
+    } catch(error) {
+      throw new Error(error);
+    }
+  }
+
+  private async authorize(credentials:AuthorizeCredentials): Promise<Oauth2Client> {
+    const clientSecret = credentials.installed.client_secret;
+    const clientId = credentials.installed.client_id;
+    const redirectUrl = credentials.installed.redirect_uris[0];
+    const auth = new GoogleAuth();
+    const oauth2Client: Oauth2Client = new auth.OAuth2(clientId, clientSecret, redirectUrl);
+
+    try {
+      const token = this.getToken();
+      oauth2Client.credentials = typeof token === 'string' ? JSON.parse(token) : token;
+      return oauth2Client;
+    } catch(error) {
+      return await this.getNewToken(oauth2Client);
+    }
+  }
+
+  private getToken(): string|object {
+    return fs.readFileSync(this.tokenPath, 'utf8');
+  }
+
+  private async getNewToken(oauth2Client:Oauth2Client): Promise<Oauth2Client> {
+    try {
+      const authUrl = oauth2Client.generateAuthUrl({
+        access_type: 'offline',
+        scope: SCOPES
+      });
+      const code: string = this.readline(authUrl);
+      const token: any = await this.getOauth2ClientToken(oauth2Client, code);
+      oauth2Client.credentials = token;
+      this.storeToken(token);
+      return oauth2Client;
+    } catch (error) {
+      throw getMessage('G_OUTH_ACCESS_ERROR',  error.message);
+    }
+  }
+
+  private readline(authUrl: string): string {
+    return readlineSync.question(getMessage('G_OUTH_ENTER_CODE', authUrl), {
+      hideEchoBack: true
+    });
+  }
+
+  private getOauth2ClientToken(oauth2Client:Oauth2Client, code: string): Promise<any> {
+    return new Promise((resolve:Function, reject:Function) => {
+      oauth2Client.getToken(code, (error:Object, token:Object) => {
+        if (error)
+          return reject(error);
+        else
+          return resolve(token);
+      });
+    });
+  }
+
+  private storeToken(token:string): void {
+    try {
+      fs.mkdirSync(this.tokenDir);
+    } catch (error) {
+      if (error.code !== EEXIST) {
+        throw error;
+      }
+    }
+    fs.writeFileSync(this.tokenPath, JSON.stringify(token));
+    console.log(getMessage('G_OUTH_STORED_TOKEN', this.tokenPath));
```

Comment:
It's nice to see what results have been appended
```
Appending...
541.09  541.099 559.483 559.416 559.416 559.483 559.483

Appended
541.09  541.099 559.483 559.416 559.416 559.483 559.483

```

Wanna get rid of it or use kinda logger?